lib/spack/docs/containers.rst

```diff
@@ -0,0 +1,308 @@
+built on `Docker Hub <https://hub.docker.com/u/spack>`_
+at every push to ``develop`` or to a release branch. The OSes that
+are currently supported are:
+
+.. _containers-supported-os:
+
+.. list-table:: Supported operating systems
+   :header-rows: 1
+
+   * - Operating System
+     - Base Image
+     - Spack Image
+   * - Ubuntu 16.04
+     - ``ubuntu:16.04``
+     - ``spack/ubuntu-xenial``
+   * - Ubuntu 18.04
+     - ``ubuntu:16.04``
+     - ``spack/ubuntu-bionic``
+   * - CentOS 6
+     - ``centos:6``
+     - ``spack/centos6``
+   * - CentOS 7
+     - ``centos:7``
+     - ``spack/centos7``
+
+All the images are tagged with the corresponding release of Spack:
+
+.. image:: dockerhub_spack.png
+
+with the exception of the ``latest`` tag that points to the HEAD of the
+``develop`` branch. These images are available for anyone to use and take
+care of all the repetitive tasks that are necessary to setup Spack within
+a container. The container recipes generated by Spack use these base
+images for their ``build`` stage.
+
+
+-------------------------
+Environment Configuration
+-------------------------
+
+A Spack Environment can be configured to generate container recipes by
+adding a ``container`` section with extra metadata metadata necessary for
+the image build:
+
+.. code-block:: yaml
+
+   spack:
+     specs:
+     - gromacs+mpi
+     - mpich
+
+     container:
+       # Select the format of the recipe e.g. docker,
+       # singularity or anything else that is currently supported
+       format: docker
+
+       # Select from a valid list of images
+       base:
+         image: "centos:7"
+         spack: develop
+
+       # Whether or not to strip binaries
+       strip: true
+
+       # Additional system packages that are needed at runtime
+       packages:
+       - libgomp
+
+       # Extra instructions
+       extra_instructions:
+         final: |
+   RUN echo 'export PS1="\[$(tput bold)\]\[$(tput setaf 1)\][gromacs]\[$(tput setaf 2)\]\u\[$(tput sgr0)\]:\w $ \[$(tput sgr0)\]"' >> ~/.bashrc
+
+       # Labels for the image
+       labels:
+         app: "gromacs"
+         mpi: "mpich"
+
+The tables below describe the supported configuration options:
+
+.. list-table:: General configuration options for the ``container`` section of ``spack.yaml``
+   :header-rows: 1
+
+   * - Option Name
+     - Description
+     - Allowed Values
+     - Required
+   * - ``format``
+     - The format of the recipe
+     - ``docker`` or ``singularity``
+     - Yes
+   * - ``base:image``
+     - Base image for ``final`` stage
+     - See :ref:`containers-supported-os`
+     - Yes
+   * - ``base:spack``
+     - Version of Spack
+     - Valid tags for ``base:image``
+     - Yes
+   * - ``strip``
+     - Whether to strip binaries
+     - ``true`` (default) or ``false``
+     - No
+   * - ``packages``
+     - System packages to be installed
+     - Valid packages for the ``final`` OS
+     - No
+   * - ``extra_instructions:build``
```

Comment:
I think this should be `extra_commands`.  Shell commands are typically called commands, and Docker constructs like `RUN` are also called "commands" throughout thee Docker docs

Reply:
That' fine with me, but this term was taken from [Docker documentation](https://docs.docker.com/engine/reference/builder/):

> Docker can build images automatically by reading the instructions from a Dockerfile

or

> To use a file in the build context, the Dockerfile refers to the file specified in an instruction, for example, a COPY instruction.

Reply:
@tgamblin I'll wait for confirmation on this. Docker documentation refers to "instruction" for things like `RUN` etc. and "command" for the command to be executed by `RUN`:

> The RUN instruction will execute any commands in a new layer on top of the current image and commit the results.

We are not tied to Docker's nomenclature though.

Reply:
ok let's leave it if this is the case -- I seem to have misread it.  If they say instructions I think we should at least try to be consistent.  Maybe the docs could say something like "extra instructions (e.g., `RUN`, `COPY`, etc.) for the generated Dockerfile"

Reply:
Done